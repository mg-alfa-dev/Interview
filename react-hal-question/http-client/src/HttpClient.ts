import type { HALDocument } from 'http-client';
import stateMachine from './state-machine';

const HttpClient = {
  async get(href: string): Promise<HALDocument> {
    console.log(`Processing request for ${href}`);

    const result = await (async () => {
      let match: RegExpExecArray | null;
      const qs = new URL(`http://dummy${href}`).searchParams;

      if (href === '/repositoryList') {
        return stateMachine.getRepositoryList();
      } else if ((match = /^\/repository\/([^/]+)\/?$/.exec(href))) {
        return stateMachine.getRepository(match[1]);
      } else if ((match = /^\/repository\/([^/]+)\/fetchBranchDetails\b/.exec(href))) {
        let id: string | null;
        if (!(id = qs.get('id'))) {
          throw new Error('Branch id must be specified.');
        }

        return stateMachine.getBranchDetails(id, match[1]);
      }

      throw new Error(`Resource not found: ${href}`);
    })();

    console.log('Result:');
    console.dir(result);
    return result;
  },

  async post(href: string, body?: any): Promise<HALDocument> {
    console.log(`Processing request for ${href}`);

    const result = await (async () => {
      let match: RegExpExecArray | null;
      const qs = new URL(`http://dummy${href}`).searchParams;

      if ((match = /^\/repository\/([^/]+)\/add-branch\b/.exec(href))) {
        const repoId = match[1];
        let name: string | null;
        if (!(name = qs.get('name'))) {
          throw new Error('Branch name must be specified.');
        }

        return stateMachine.createBranch(repoId, name);
      }

      throw new Error(`Resource not found: ${href}`);
    })();

    console.log('Result:');
    console.dir(result);
    return result;
  },

  __reset() {
    console.debug('Resetting state');
    stateMachine.reset();
  },
};

export default HttpClient;
