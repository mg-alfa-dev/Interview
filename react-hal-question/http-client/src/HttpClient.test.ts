import HttpClient from './HttpClient';
import stateMachine from './state-machine';

jest.mock('./state-machine');

afterEach(jest.resetAllMocks);

const goodURLs: [keyof typeof HttpClient, string, keyof typeof import('./state-machine').default, string, any[]][] = [
  ['get', 'fetch the repository list', 'getRepositoryList', '/repositoryList', []],
  ['get', 'fetch a repository', 'getRepository', '/repository/abcd', ['abcd']],
  ['get', 'fetch branch details', 'getBranchDetails', '/repository/abcd/fetchBranchDetails?id=1234', ['1234', 'abcd']],
  ['post', 'add a branch', 'createBranch', '/repository/abcd/add-branch/?name=foo', ['abcd', 'foo']],
];

test.each(goodURLs)('calling HttpClient.%s() to %s calls stateMachine.%s', async (clientMethod, _, stateMachineMethod, uri, args) => {
  await HttpClient[clientMethod](uri);
  expect(stateMachine[stateMachineMethod]).toHaveBeenCalledTimes(1);
  expect(stateMachine[stateMachineMethod]).toHaveBeenCalledWith(...args);
});

const badURLs: [keyof typeof HttpClient, string, string][] = [
  ['get', 'with a bad url', '/aslkdjf'],
  ['get', 'to add a branch', '/repository/abcd/add-branch?name=foo'],
  ['get', 'to get a repository without an id', '/repository'],
  ['get', 'to get a branch without an id', '/repository/abcd/fetchBranchDetails'],
  ['post', 'with a bad url', '/aslkdjf'],
  ['post', 'to add a branch without a name', '/repository/abcd/add-branch'],
  ['post', 'to add a branch with an empty name', '/repository/abcd/add-branch?name='],
  ['post', 'to fetch the repository list', '/repositoryList'],
  ['post', 'to fetch a repository', '/repository/abcd'],
  ['post', 'to fetch branch details', '/repository/abcd/fetchBranchDetails?id=1234'],
];

test.each(badURLs)('calling HttpClient.%s() %s throws an error', async (method, _, uri) => {
  await expect(HttpClient[method](uri)).rejects.toMatchSnapshot();
});
