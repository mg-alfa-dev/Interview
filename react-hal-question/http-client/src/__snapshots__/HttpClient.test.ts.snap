// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling HttpClient.get() to add a branch throws an error 1`] = `[Error: Resource not found: /repository/abcd/add-branch?name=foo]`;

exports[`calling HttpClient.get() to get a branch without an id throws an error 1`] = `[Error: Branch id must be specified.]`;

exports[`calling HttpClient.get() to get a repository without an id throws an error 1`] = `[Error: Resource not found: /repository]`;

exports[`calling HttpClient.get() with a bad url throws an error 1`] = `[Error: Resource not found: /aslkdjf]`;

exports[`calling HttpClient.post() to add a branch with an empty name throws an error 1`] = `[Error: Branch name must be specified.]`;

exports[`calling HttpClient.post() to add a branch without a name throws an error 1`] = `[Error: Branch name must be specified.]`;

exports[`calling HttpClient.post() to fetch a repository throws an error 1`] = `[Error: Resource not found: /repository/abcd]`;

exports[`calling HttpClient.post() to fetch branch details throws an error 1`] = `[Error: Resource not found: /repository/abcd/fetchBranchDetails?id=1234]`;

exports[`calling HttpClient.post() to fetch the repository list throws an error 1`] = `[Error: Resource not found: /repositoryList]`;

exports[`calling HttpClient.post() with a bad url throws an error 1`] = `[Error: Resource not found: /aslkdjf]`;
